Add "record.version"; add "record.loopback" for LoopBack applications

bajtos opened the pull request:
connected to strongloop-internal/scrum-nodeops#422

**Add `record.version` property**

Add a new record property `version` containing the version of
strong-express-metrics that produced the record. This allows consumers
to detect different record formats based on producer's version.

**Fill `record.loopback` for LoopBack applications**

When the metrics middleware is added to a LoopBack application
(or a strong-remoting application), a new top-level property "loopback"
is added to the metrics record. This property contains an object
with the following properties:
- modelName
  The name of the LoopBack model (shared class).
- remoteMethod
  The name of the remote method invoked, e.g. "find" or "prototype.updateAttributes".
- instanceId
  The ID of the target instance (when available).

Examples:

```
POST /api/Cars
  modelName: Car
  remoteMethod: create
  (instanceId not available)

PUT /api/Cars/123
  modelName: Car
  remoteMethod: prototype.updateOrCreate
  instanceId: 123

GET /api/Cars/123
  modelName: Car
  remoteMethod: findById
  instanceId: 123
```

Connect to strongloop/loopback#1276
Requires https://github.com/strongloop/strong-remoting/pull/199

/to @sam-github please review
/to @kraman @altsang please check that  the metadata meets your requirements
/cc @ritch you may want to take a look too

Commit:
Fill `record.loopback` for LoopBack applications

When the metrics middleware is added to a LoopBack application
(or a strong-remoting application), a new top-level property `loopback`
is added to the metrics record. This property contains an object
with the following properties:

  modelName
    The name of the LoopBack model (shared class).

  remoteMethod
    The name of the remote method invoked, e.g.
    "find" or "prototype.updateAttributes".

  instanceId
    The ID of the target instance (when available).

Examples:

    POST /api/Cars
      modelName: Car
      remoteMethod: create
      (instanceId not available)

    PUT /api/Cars/123
      modelName: Car
      remoteMethod: prototype.updateOrCreate
      instanceId: 123

    GET /api/Cars/123
      modelName: Car
      remoteMethod: findById
      instanceId: 123

bajtos commented:
Here is an example showing a full metrics record produced for a LoopBack application:

``` json
{
  "timestamp": 1428565224361,
  "client": {
    "address": "127.0.0.1"
  },
  "request": {
    "method": "GET",
    "url": "/cars/1234"
  },
  "response": {
    "status": 404,
    "duration": 6
  },
  "process": {
    "pid": 55941
  },
  "data": {
    "loopback": {
      "remoteMethod": "findById",
      "modelName": "Car",
      "instanceId": 1234
    }
  }
}
```

sam-github commented:
Why not put `loopback` at the top-level? Wasn't `data` meant to be the only safe place users could put their own data?

LGTM otherwise

altsang commented:
I assume to keep it more decoupled as truly an express middleware with no inferred dependency or assocation to LB directly

bajtos commented:
> @sam-github Why not put `loopback` at the top-level? Wasn't `data` meant to be the only safe place users could put their own data?

I have considered that option too. My view is that anything that is not express-related should go to `data`. That way we can add support for other frameworks later, e.g. `data.sailsjs` or `data.krakenjs` without the need to change the record structure.

It also creates a sort of a convention: put your data inside `data.{namespace}` object.

I can be convinced to move the `loopback` object to the root level, if you think it's a better solution.

> I assume to keep it more decoupled as truly an express middleware with no inferred dependency or association to LB directly

Exactly.

altsang commented:
​makes sense +1​

On Thu, Apr 9, 2015 at 9:06 AM, Miroslav Bajtoš notifications@github.com
wrote:

> @sam-github https://github.com/sam-github Why not put loopback at the
> top-level? Wasn't data meant to be the only safe place users could put
> their own data?
> 
> I have considered that option too. My view is that anything that is not
> express-related should go to data. That way we can add support for other
> frameworks later, e.g. data.sailsjs or data.krakenjs without the need to
> change the record structure.
> 
> It also creates a sort of a convention: put your data inside
> data.{namespace} object.
> 
> I can be convinced to move the loopback object to the root level, if you
> think it's a better solution.
> 
> I assume to keep it more decoupled as truly an express middleware with no
> inferred dependency or association to LB directly
> 
> Exactly.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/strongloop/strong-express-metrics/pull/3#issuecomment-91275499
> .

## 

Al Tsang
Co-Founder and CTO @ StrongLoop, Inc.

_StrongLoop http://strongloop.com/_\* makes it easy to **develop APIs
http://strongloop.com/mobile-application-development/loopback/** in Node,
plus get DevOps capabilities
http://strongloop.com/node-js-performance/strongops/ like monitoring,
debugging and clustering*_._

bajtos commented:
@slnode test please

bajtos commented:
I guess we could move all loopback stuff under `request` too:

``` json
{
  "timestamp": 1428565224361,
  "request": {
    "method": "GET",
    "url": "/cars/1234",
    "remoteMethod": "findById",
    "modelName": "Car",
    "instanceId": 1234
  },
  "etc.": "..."
}
```

anthonyettinger commented:
@bajtos that last example there looks easier to work with.

sam-github commented:
The thing about record.fu. and record.data.fu, is that it introduces two namespaces, and it must be clear why.

Why should the top-level be for "express", and everything else go in "data"? Even if we were likely to want to add a new top-level express-specific property (seems unlikely), we could just choose a property name that doesn't conflict. And we wouldn't call it `loopback`! So its safe for us to add properties at the top-level - we aren't going to step on our own toes, so I don't see why we should add unnecessary nesting, expecially under such a general name as "data" (everything in the record is data of some sort).

I think it makes more sense for the top-level to be properties directly implemented by strong-express-metrics, and the `record.data` to be for users to do with as they wish, safe in the knowledge that they will never be overwriting anything we decided to add (like hapi-specific info).

If we are putting loopback into `record.data`, then they no longer have a safe place to extend the record. So then we'd need a `record.user` or something... and it all gets too convoluted, IMO.

I don't have strong feelings about whether the new records go into record.request, or whether that's too loopback specific. Personally, I'd put it into record.loopback for now, the data (like "model") looks very loopback specific, and also add a `record.version` field which is the version of strong-express-metrics, so we can extend or change the record later, if we need to.

sam-github commented:
Btw, the original intention of `data` being for the user, not us, was documented: https://github.com/strongloop/strong-express-metrics#record-format

Commit:
Add `record.version` property

Add a new record property `version` containing the version of
strong-express-metrics that produced the record. This allows consumers
to detect different record formats based on producer's version.

bajtos commented:
@sam-github reworked per your suggestions, we have `record.loopback` and `record.version` now. I have also updated README to describe these additions. PTAL again, feel free to merge the patch via the green button if all LGTY now.

sam-github commented:
LGTM

bajtos commented:
Released as `strong-express-metrics@1.1.0`

/cc @kraman @anthonyettinger
